internal/cloud/realm/client.go

```diff
@@ -41,6 +57,14 @@ func (c *client) do(method, path string, options api.RequestOptions) (*http.Resp
 	}
 	req.Header.Set(requestOriginHeader, cliHeaderValue)
 
+	// TODO REALMC-7156 How do we want to handle refresh tokens here
+	if c.session != nil {
+		if len(c.session.AccessToken) == 0 {
+			return nil, errors.New("the current session is invalid. This can happen if you have not yet logged in or if your refresh token has expired")
+		}
+		req.Header.Add("Authorization", "Bearer "+c.session.AccessToken)
+	}
+
```

Comment:
This logic is pretty 'dumb'. We have no idea if the session here is valid or not and don't really check anything here. This means that even though we could know that a future request will fail, we do nothing and let that happen. Open to some ideas/discussions here.

Reply:
OK, so I like where this is headed, but I think we could make a few tweaks here to set us up nicely in the future.

first change will be to the `api.RequestOptions` struct that this `do` function takes in as an arg:
1. Remove the `Header` field
1. Add `ContentType string`, `UseAuth bool`, and `RefreshAuth bool` fields
1. We can also remove the `JSONRequestOptions` func, we'll handle that elsewhere now

now in `client.go`, we're going to have two private methods responsible for executing requests:
`func (c *client) do(method, path string, options api.RequestOptions) (*http.Response, error)`
`func (c *client) doJSON(method, path string, payload interface{}, options api.AuthOptions) (*http.Response, error)`

I'll leave the implementations to you, but it stands to reason `doJSON` would be doing the stuff you saw in `JSONRequestOptions` and then just returning `c.do(...)`

So you'll have to tweak some of the implementation in the `do` function to accommodate the new options (with content type, make sure you only `Set` it on the header if its not blank).  For authorization, I think it'd make sense to have the following helper function:
`func (c *client) getAuth(options api.RequestOptions) (string, error)`
which was responsible for returning the appropriate session token based on the `bool`s found in `RequetOptions`.  It is here also that we'd return an error if say `UseAuth` was `true` but then `Session.AccessToken` was blank

thoughts?

Reply:
Discussed offline, done!